feat(db): update devices to match new requirements

philbooth opened the pull request:
There are ~~two~~ three changes here:

1. Fixes #114, adding a column for the push callback public key.

2. [Return an error if a device attempts to register a session token that is already registered to another device](https://github.com/mozilla/fxa/tree/master/features/FxA-45-device-registration-api#registered-device-1).

3. [Return an error from the `POST` and `DELETE` endpoints if an unrecognised device id is received](https://github.com/mozilla/fxa/issues/78).

I put them ~~both~~ all into the same commit because I wanted to limit database changes to a single migration and I figured one migration should always be one atomic commit.

~~My approach to `2`, especially, may be (is probably?) wrong. It performs two extra queries in the worst case and then aborts the stored procedure using `SIGNAL SQLSTATE`~~. I look forward to the next stage of my on-the-job databases education in your comments. :grin:

@dannycoates && @rfk, r?

rfk commented on fxa-auth-db-server/test/backend/db_tests.js:
```diff
@@ -832,13 +839,26 @@ module.exports = function(config, DB) {
+                  type: null,
+                  callbackURL: '',
+                  callbackPublicKey: zeroBuffer32
+                })
```
IIRC, part of what this is testing is sensible handling of partial updates, so let's be careful we don't lose that by specifying all the extra device info here.

rfk commented on fxa-auth-db-server/lib/error.js:
```diff
@@ -50,6 +50,17 @@ AppError.incorrectPassword = function () {
+      code: 400,
+      error: 'Bad request',
+      errno: 104,
+      message: 'Session token is registered with a different device'
```
Something about this naming seems a bit backward to me; in my head it's the *device* that's being registered, not the session token.  Maybe something more generic like "deviceIdConflict"?  Not sure...

philbooth replied:
`deviceSessionConflict`?

rfk commented on fxa-auth-db-server/test/backend/db_tests.js:
```diff
@@ -849,15 +869,67 @@ module.exports = function(config, DB) {
+                  callbackPublicKey: hex32()
+                })
+                .then(function() {
+                  t.fail('adding a second device does not fail when the session token is already registered')
```
A query about wording here.  Every time I read these, I parse them as statements of desired behaviour.  So I just read the above, thought "adding a second device does not fail when already registered? that's doesn't sound right..." and took a minute to realize that it was *failing* if that was the case.  What do you think about writing these failure guards in past tense, so it's more obvious that they're telling you that something went wrong? 

e.g. "adding a second device should have failed when..."

rfk commented on lib/db/mysql.js:
```diff
@@ -283,9 +283,15 @@ module.exports = function (log, error) {
       ]
-    )
+    ).catch(function (err) {
+      if (err.code === 500 && err.errno === 1644) {
```
This `1644` is a magic constant, we should name it something meaningful.

rfk commented:
> aborts the stored procedure using SIGNAL SQLSTATE

I'll be honest, we're well beyond my familiarity with stored procedures here :-)

IIUC the desired behaviour is:
* If the sessionToken exists in "devices" table for a different deviceid, return an error
* If the sessionToken doesn't match what's in the "devices" table for this device, update the stored value.

We could make the db enforce this by creating a unique constraint on (uid, sessionTokenId), so that attempts to insert a duplicate sessionTokenId would fail.  This would leave us with a single db operation in the case of well-behaved devices, but it might interfere with our use of ON DUPLICATE KEY UPDATE to handle the upsert.

rfk commented:
(We may also find that, as we map out the error codes and edge cases, it becomes cleaner to use separate "insert new device" and "update existing device" operations rather than a single upsert).

philbooth commented:
> (We may also find that, as we map out the error codes and edge cases, it becomes cleaner to use separate "insert new device" and "update existing device" operations rather than a single upsert).

Doing that would make it easier to address https://github.com/mozilla/fxa/issues/78 I think.

philbooth commented on lib/db/schema/patch-019-020.sql:
```diff
@@ -0,0 +1,95 @@
+    d.uid = inUid;
+END;
+
+DROP PROCEDURE `upsertDevice_1`;
```
I figure this is safe to do because it's never been used...?

rfk replied:
Should be safe, but but it might make e.g. rollbacks harder.  I suggest just leaving it in place and we'll garbage-collect alongside other obsolete stored procedures.

philbooth commented:
@rfk, thanks for your feedback. I've pushed fixes for all of your comments, including the switch to using a unique constraint to guarantee session token sanity.

I also followed your suggestion of splitting the upsert into two queries. Doing that enabled me to address https://github.com/mozilla/fxa/issues/78, returning an error when an unrecognised device id is received in the POST or DELETE endpoint.

philbooth commented on lib/db/mysql.js:
```diff
@@ -283,8 +283,52 @@ module.exports = function (log, error) {
+      null,
+      function (err) {
+        if (err.errno === ER_DUP_ENTRY) {
+          if (err.toString().match(/ for key 'UQ_devices_sessionTokenId'$/)) {
```
This is pretty flakey, I realise, but it's the only way I could find to distinguish the two errors. But maybe distingushing between them is unnecessary and it's fine if they both return the same `errno`? I'd *love* to get rid of this if that is the case. :smile:

philbooth replied:
Well this is embarassing. It turns out that this requirement was entirely manufactured inside my own imagination. I am deleting the weirdness.

philbooth commented:
@rfk, just in case you looked at this on Friday, I should flag up that I've pushed a minor change since then.

There was a discrepancy between how the memory store and MySQL were treating `callbackPublicKey`, because the memory store wasn't coercing it to a buffer in the empty string case. I added some extra tests to `test/local/db_tests.js` and fixed the issue itself inside `updateDeviceRecord` in `lib/db/mem.js`.

rfk commented on fxa-auth-db-server/docs/API.md:
```diff
@@ -59,7 +59,8 @@ The following datatypes are used throughout this document:
 * Devices:
-    * upsertDevice              : `PUT /account/:id/device/:deviceId`
+    * createDevice              : `PUT /account/:id/device/:deviceId`
+    * updateDevice              : `POST /account/:id/device/:deviceId`
```
I'm not thrilled about using `PUT` and `POST` on the same end-point with such different semantics.  We seem to have a convention of using `PUT /blah/:id` for creation and `POST /blah/:id/update` for updates on e.g. some of the token types, which we should consider just copying here for consistency.

rfk commented on fxa-auth-db-server/docs/API.md:
```diff
@@ -680,6 +683,56 @@ Content-Type: application/json
 * Status Code : 200 OK
     * Content-Type : 'application/json'
     * Body : `[ ... <see example> ...]`
+* Status Code : 409 Conflict
```
`409` is a good code for this from a restfulness standpoint, but I think it would be out-of-place in the auth-server API which tends to use a plain `400` for just about everything that's not specifically covered by a broad `4XX` code convention.  API consumers may find themselves wondering why other types of error don't also generate a `409`.

rfk replied:
Huh, so I see in https://github.com/mozilla/fxa-auth-server/pull/1112/ that we're already using `409` here for other requests, and translating it into 400 errors in the auth-server.  So this is all good, carry on :-)

rfk commented:
API nits aside, this LGTM.

philbooth commented:
Thanks for the feedback again @rfk, fixes pushed.
